framework/include/constraints/NodeElemConstraint.h

```diff
@@ -0,0 +1,254 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef NODEELEMCONSTRAINT_H
+#define NODEELEMCONSTRAINT_H
+
+// MOOSE includes
+#include "Constraint.h"
+#include "NeighborCoupleableMooseVariableDependencyIntermediateInterface.h"
+
+// Forward Declarations
+class NodeElemConstraint;
+
+// libMesh forward declarations
+namespace libMesh
+{
+template <typename T>
+class SparseMatrix;
+}
+
+template <>
+InputParameters validParams<NodeElemConstraint>();
+
+/**
+ * A NodeElemConstraint is used when you need to create constraints between
+ * a slave node and a master element. It works by allowing you to modify the
+ * residual and jacobian entries on the slave node and the master element.
+ */
+class NodeElemConstraint : public Constraint,
+                           public NeighborCoupleableMooseVariableDependencyIntermediateInterface,
+                           public NeighborMooseVariableInterface<Real>
+{
+public:
+  NodeElemConstraint(const InputParameters & parameters);
+  virtual ~NodeElemConstraint();
+
+  /// Compute the value the slave node should have at the beginning of a timestep.
+  void computeSlaveValue(NumericVector<Number> & current_solution);
+
+  /// Computes the residual Nodal residual.
+  virtual void computeResidual();
+
+  /// Computes the jacobian for the current element.
+  virtual void computeJacobian();
+
+  /// Computes d-residual / d-jvar...
+  virtual void computeOffDiagJacobian(unsigned int jvar);
+
+  /// Gets the indices for all dofs connected to the constraint
+  virtual void getConnectedDofIndices(unsigned int var_num);
+
+  /**
+   * Whether or not this constraint should be applied.
+   * @return bool true if this constraint is active, false otherwise
+   */
+  virtual bool shouldApply() { return true; }
+
+  /// Compute reaction force of the constraint depending on the formulation
+  virtual void computeConstraintForce(){};
```

Comment:
A couple of things about this:

1.  The name `computeConstraintForce()` is specific to mechanics. Both this base class and the derived class in MOOSE are very generic classes that are applicable to any physics, so this needs to be named accordingly.
2.  I'm not convinced that this method even needs to exist in the base class. It might make sense (with a different name) to define as a protected member of the derived class, but I don't think we really need to call it from NonlinearSystemBase.